### PR TITLE
In some rare cases, invocation can be deleted and not nulled while we…

### PR DESCRIPTION
**Pull request by Civil**
… are upgrading locks

Related to #193 but fixes only crash.

### CODE REVIEW

**Comment by grobian** [comment]
This was what I did before, the problem was that this reduces the capacity of the aggregator even more, which gave some issues in the existing setup back then.

**Comment by Civil** [comment]
Yeah, but without that it's crashing because of the race between 'aggregator_putmetric' and 'aggregator_expire'.

**Comment by grobian** [comment]
Yes, I'm willing to push this, was just mentioning it may not be what you want. I'm thinking maybe we can do a cheaper check for the object to exist when the writelock has been acquired.
However, that section is too much shared anyway, in the long term plan, it needs splitting so multiple threads can concurrently do some work together.  So if you want this fix, I'll merge it.

File: aggregator.c [hunk 1]
```diff
@@ -237,18 +237,10 @@ aggregator_putmetric(
 			if (invocation->hash == omhash && \
 					strcmp(o, invocation->metric) == 0)  /* match */ \
 				break;
-		pthread_rwlock_rdlock(&compute->invlock);
+		pthread_rwlock_wrlock(&compute->invlock);
 		find_invocation(ometric);
 		/* switch to a write lock from here, since we've found what we
 		 * were looking for (or are going to create it) and modify it */
```
Comment (grobian) [inline aggregator.c]: if you start with a write-lock, then this comment should go as well, since we aint switching any more.

**Comment by Civil** [comment]
Well, I also don't know yet why it stops sending data out. My idea right now is to try to make it slower but stricter at thread handling by fixing TSAN issues and see if it eventually goes away. If not - look somewhere else. After that we can try to improve the performance to a level that's good enough for it to work until this is properly addressed...

File: aggregator.c (hunk 1 above)
Comment (Civil) [inline aggregator.c]: thx, pushed fixed commit

**Pull request merged**